raft-compat/prevote_test.go

```diff
@@ -0,0 +1,294 @@
+				//Create an upgraded node with the store
+				rUIT := testcluster.InitUITWithStore(t, id, store.(*raftprevious.InmemStore), func(config *raft.Config) {
+					config.PreVote = tc.preVote
+				})
+				future := getLeader.GetRaft().(*raftprevious.Raft).AddVoter(raftprevious.ServerID(rUIT.GetLocalID()), raftprevious.ServerAddress(rUIT.GetLocalAddr()), 0, 0)
+				utils.WaitFuture(t, future)
+				//Add the new node to the cluster
+				cluster.AddNode(rUIT)
+
+				// Wait enough to have the configuration propagated.
+				time.Sleep(time.Second)
+
+				//Apply some logs
+				future = getLeader.GetRaft().(*raftprevious.Raft).Apply([]byte("test2"), time.Second)
+				require.NoError(t, future.Error())
+
+				// Check Leader haven't changed as we haven't replaced the leader yet
+				a, _ = getLeader.GetRaft().(*raftprevious.Raft).LeaderWithID()
+				require.Equal(t, a, leader)
+			}
+			// keep a reference to the store
+			store := cluster.Store(getLeader.GetLocalID())
+			id := getLeader.GetLocalID()
+
+			//Remove and shutdown the leader node
+			tc.Leave(t, cluster, getLeader.GetLocalID())
+
+			// Delete the old leader node from the cluster
+			cluster.DeleteNode(getLeader.GetLocalID())
+			oldLeaderID := getLeader.GetLocalID()
+
+			// Wait for a new leader to be elected
+			utils.WaitForNewLeader(t, oldLeaderID, cluster)
+			getLeader = cluster.GetLeader()
+			require.NotEmpty(t, getLeader)
+
+			// Create a new node to replace the deleted one
+			rUIT := testcluster.InitUITWithStore(t, id, store.(*raftprevious.InmemStore), func(config *raft.Config) { config.PreVote = true })
+			fa := getLeader.GetRaft().(*raft.Raft).AddVoter(raft.ServerID(rUIT.GetLocalID()), raft.ServerAddress(rUIT.GetLocalAddr()), 0, 0)
+			utils.WaitFuture(t, fa)
+
+			// Wait for new leader, (this happens because of not having prevote)
+			utils.WaitForNewLeader(t, "", cluster)
+			newLeaderID := rUIT.GetLeaderID()
+			require.NotEmpty(t, newLeaderID)
+
+			require.NotEqual(t, newLeaderID, leader)
+
+			newLeader := cluster.GetLeader()
+			//Apply some logs
+			future = newLeader.GetRaft().(*raft.Raft).Apply([]byte("test2"), time.Second)
+			require.NoError(t, future.Error())
+
+			// Check Leader haven't changed as we haven't replaced the leader yet
+			newAddr, _ := newLeader.GetRaft().(*raft.Raft).LeaderWithID()
+			require.Equal(t, string(newAddr), newLeader.GetLocalAddr())
+
+			require.Equal(t, tc.numNodes, rUIT.NumLogs())
+		})
+	}
+
+}
+
+func TestRaft_PreVote_Rollback(t *testing.T) {
+	leaveTransfer := func(t *testing.T, cluster testcluster.RaftCluster, id string) {
+		if cluster.GetLeader().GetLocalID() == id {
+			transfer := cluster.Raft(id).(*raft.Raft).LeadershipTransfer()
+			utils.WaitFuture(t, transfer)
+		}
+		f := cluster.Raft(id).(*raft.Raft).Shutdown()
+		utils.WaitFuture(t, f)
+	}
+	leaveNoTransfer := func(t *testing.T, cluster testcluster.RaftCluster, id string) {
+		fr := cluster.GetLeader().GetRaft().(*raft.Raft).RemoveServer(raft.ServerID(id), 0, 0)
```

Comment:
I'm not sure I understand 🤔. This will trigger prevote, specifically when we get to upgrade the leader but it won't test term inflation. Prevote happen in all leader election scenarios before a real vote, so in this case, when we upgrade the leader node (the last node to be upgraded), a leader election will happen and it will involve a pre-vote.

Reply:
Right it will trigger the pre-vote when the leader gets rolled but not when followers are being shut down.